File: services/chatgpt_core/gopay_phone.py
from __future__ import annotations

import re
from typing import Any

DEFAULT_GOPAY_PHONE_COUNTRY_CODE = "62"
GOPAY_RECOGNIZED_COUNTRY_CODES_KEY = "chatgpt_gopay_recognized_country_codes"


def digits_only(value: Any) -> str:
    return re.sub(r"\D", "", str(value or ""))


def normalize_gopay_recognized_country_codes(value: Any) -> list[str]:
    if isinstance(value, (list, tuple, set)):
        raw_items = [digits_only(item) for item in value]
    else:
        raw_items = re.findall(r"\d+", str(value or ""))

    normalized: list[str] = []
    seen: set[str] = set()
    for item in raw_items:
        code = digits_only(item)
        if not code or code in seen:
            continue
        seen.add(code)
        normalized.append(code)

    return normalized or [DEFAULT_GOPAY_PHONE_COUNTRY_CODE]


def split_gopay_phone_input(
    phone_country_code: Any,
    phone_number: Any,
    recognized_country_codes: Any,
    *,
    default_country_code: str = DEFAULT_GOPAY_PHONE_COUNTRY_CODE,
) -> dict[str, str]:
    country_code = digits_only(phone_country_code) or digits_only(default_country_code)
    number = digits_only(phone_number)
    codes = normalize_gopay_recognized_country_codes(recognized_country_codes)

    for code in sorted(codes, key=lambda item: (-len(item), codes.index(item))):
        if number.startswith(code) and len(number) > len(code):
            return {
                "phone_country_code": code,
                "phone_number": number[len(code):],
            }

    return {
        "phone_country_code": country_code or DEFAULT_GOPAY_PHONE_COUNTRY_CODE,
        "phone_number": number,
    }

File: frontend/src/lib/gopayPhone.ts
export const DEFAULT_GOPAY_PHONE_COUNTRY_CODE = '62'

export function normalizeGopayPhonePart(value: unknown) {
  return String(value || '').replace(/\D/g, '')
}

export function normalizeGopayRecognizedCountryCodes(value: unknown): string[] {
  const rawItems = Array.isArray(value)
    ? value.map((item) => normalizeGopayPhonePart(item))
    : Array.from(String(value || '').matchAll(/\d+/g)).map((match) => match[0])
  const seen = new Set<string>()
  const normalized: string[] = []
  rawItems.forEach((item) => {
    const code = normalizeGopayPhonePart(item)
    if (!code || seen.has(code)) return
    seen.add(code)
    normalized.push(code)
  })
  return normalized.length > 0 ? normalized : [DEFAULT_GOPAY_PHONE_COUNTRY_CODE]
}

export function splitGopayPhoneInput(
  phoneCountryCode: unknown,
  phoneNumber: unknown,
  recognizedCountryCodes: unknown,
) {
  const countryCode = normalizeGopayPhonePart(phoneCountryCode) || DEFAULT_GOPAY_PHONE_COUNTRY_CODE
  const number = normalizeGopayPhonePart(phoneNumber)
  const codes = normalizeGopayRecognizedCountryCodes(recognizedCountryCodes)
  const sortedCodes = codes
    .map((code, index) => ({ code, index }))
    .sort((a, b) => b.code.length - a.code.length || a.index - b.index)

  for (const { code } of sortedCodes) {
    if (number.startsWith(code) && number.length > code.length) {
      return {
        phone_country_code: code,
        phone_number: number.slice(code.length),
      }
    }
  }

  return {
    phone_country_code: countryCode,
    phone_number: number,
  }
}
